README.md
# Rusty Pickaxe

Multithreaded CPU and GPU (CUDA) miner for [Provably Rare Gems](https://gems.alphafinance.io/#/rarity), written in Rust.

There is donation on every found gem. 0.1 FTM for rarity gems and 0.001 ETH for 
loot. You can disable it in code if you think you do not owe me anything.

## Config

Same `config.json` as in [ramen](https://github.com/dmptrluke/ramen) miner. 
`loop` option works since v0.0.2. `threads` options works and is required since v0.0.4.
You will probably want to set `threads` equal or less than your CPUs thread count.
If you want to mine with cuda set `threads` to anything, it does not matter.

## Binaries (untested)

Grab one from [releases page](https://github.com/Ledarium/rusty_pickaxe/releases).
Run it like `rusty_pickaxe.exe config.json`

## Build from source
### Prerequisites

If following steps do not work for you, you can always ask for support in [Discord](https://discord.gg/xDk6enpGnM).

#### Ubuntu

For cuda install refer to https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html

Install rust: https://www.rust-lang.org/tools/install

Install extra bits: `sudo apt install build-essential libssl-dev pkg-config`

#### Windows

For cuda install refer to https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html

Install VS build tools: https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019 ,
you want "Build Tools for Visual Studio 2019". When the installer asks what parts of 
"Build Tools for Visual Studio 2019" you want, go to the "individual components" tab
and pick "x64/x86 build tools (latest)".

Install rust: https://www.rust-lang.org/tools/install

#### CUDA version

If you are getting errors like this:
```
cargo:warning=nvcc fatal   : Unsupported gpu architecture 'compute_86'
```
This means your GPU is not enabled in config. I did enable most popular ones like 
9XX/10XX, RTX 20XX/30XX. But fully enabling all of them does not make sense and could
negatively affect performance. If you are having this issue, see your `compute_XX` and
`sm_XX` params 
[here](https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/)
and write them replacing predefined ones in `src/build.rs`.

### Run

CPU: `cargo run --release config.json`

CUDA: `cargo run --features cuda --release config.json`

Only one GPU in system is supported for now. However, you can launch multiple miners
changing `N` for every card in system.

- Linux: `CUDA_VISIBLE_DEVICES=N cargo ...`
- Windows (untested): `set CUDA_VISIBLE_DEVICES=N & cargo ...`

## Hashrate test

Data is stated in MH/s.
You can share your results to appear here.

CPU | Single core | Threads | Multi core
--- | --- | --- | ---
Ryzen 3700X | 2.67 | 16 | ~18.70
Ryzen 2600 | 2.20 | 12 | ~13.44
i5-9600k@5GHz | 2.82 | 6 | ~16.92
i7-9700@3GHz | 2.08 | 8 | ~16.64
i7-10700k@5GHz | 2.50 | 8 | ~20.00

GPU | Hashrate
--- | ---
3090 | 1700
3070 | 1007
3060 Ti | 1007
2080 Ti | 1200
1080 Ti | 787
1060 | 280...324
1050 Ti | 160...183
1050 | 154
1030 | 89
 
## Acknowledgments

- Made it public thanks to generous `wenl#6575`.
- `Kakapo#5409` made nice json format for config file and wrote install instructions.
- Also thanks to `Spiry#6864`, `spiz0r#7566`, `shock#9999` for donations.
- Anonymous donations were also a big help.

Kudos to all of you! Donations are accepted here: [Ethereum](https://etherscan.io/address/0x8dd47bf52589cf12ff4703951c619821cf794b77), [Fantom](https://ftmscan.com/address/0x8dd47bf52589cf12ff4703951c619821cf794b77), [BSC](https://bscscan.com/address/0x8dd47bf52589cf12ff4703951c619821cf794b77).

## Contacts

- Discord `Booba#1974`
- Telegram @amazing_booba

## Legal notice

SOFTWARE IS PROVIDED AS IS AND AUTHOR IS NOT RESPONSIBLE FOR ANYTHING.


CHANGELOG.md
## Unreleased

### Added

## [0.0.5] - 2021-10-13

- CUDA
- Better hashrate reporting
- Job polling
- Gem count
- Dev donation

## [0.0.4] - 2021-09-22

### Added
- Multithreading, and it works

### Fixed
- Found TX is now displayed as address

## [0.0.3] - 2021-09-22

### Fixed
- Mining is not sequential now
- Poll work params every N iterations to prevent stale salts
- Refactor code to prepare for proper multithreading

## [0.0.2] - 2021-09-20

### Fixed
- `loop` option works now
- Groom code

### Added
- Donators section in readme

## [0.0.1] - 2021-09-19

- Pickaxe goes BRRRRR

### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security



test_data.py
from Crypto.Hash import keccak
from eth_abi.packed import encode_abi_packed

abi_types = ["uint256", "bytes32", "address", "address", "uint", "uint", "uint"]
packed = encode_abi_packed(
            abi_types,
            [1, b'b'*32, "0xFFcf8FDEE72ac11b5c542428B35EEF5769C409f0", "0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1", 1, 20, 2])
print(', '.join(str(byte) for byte in packed))
print(packed.hex())
k = keccak.new(digest_bits=256)
k.update(packed)
hx = k.hexdigest()
print(hx)


build.rs
#[cfg(feature = "cuda")]
extern crate cc;

#[cfg(feature = "cuda")]
static CUDA_PATH: &str = "/opt/cuda/include";

#[cfg(feature = "cuda")]
fn main() {
    built::write_built_file().expect("Failed to acquire build-time information");
    cc::Build::new()
        .cuda(true)
        .flag("-cudart=shared")
        .flag("-gencode")
        .flag("arch=compute_50,code=sm_50") // GTX 7**
        .flag("-gencode")
        .flag("arch=compute_52,code=sm_52") // GTX 9**
        .flag("-gencode")
        .flag("arch=compute_60,code=sm_60") // P100
        .flag("-gencode")
        .flag("arch=compute_61,code=sm_61") // GTX 10**
        .flag("-gencode")
        .flag("arch=compute_75,code=sm_75") // RTX 20**, 1660 Ti
        .flag("-gencode")
        .flag("arch=compute_86,code=sm_86") // RTX 30**
        .include("./src")
        .files(&["./src/keccak.cu"])
        .compile("libkeccak.a");
    println!("cargo:rustc-link-search=all={}", CUDA_PATH);
    println!("cargo:rustc-env=LD_LIBRARY_PATH={}", CUDA_PATH);
    println!("cargo:rustc-link-lib=dylib=cudart");
}

#[cfg(not(feature = "cuda"))]
fn main() {
    built::write_built_file().expect("Failed to acquire build-time information");
}


src/cpu.rs
use crate::utils::{serialize_work, Work};
use log::debug;
use tiny_keccak::{Hasher, Keccak};

pub fn prepare_data(work: &Work) -> Keccak {
    let mut h = Keccak::v256();
    h.update(&serialize_work(&work.first_block));
    //debug!("Keccak hex data: {}", String::from(bytes.to_hex()));
    h.update(&serialize_work(&work.second_block)[0..56]);
    return h;
}

pub fn optimized_hash(mut h: Keccak, salt: u64) -> [u8; 32] {
    h.update(&salt.to_be_bytes());
    let mut res = [0u8; 32];
    h.finalize(&mut res);
    return res;
}

pub fn simple_hash(work: &Work) -> [u8; 32] {
    let mut h = Keccak::v256();
    h.update(&serialize_work(&work.first_block));
    h.update(&serialize_work(&work.second_block)[0..64]);
    let mut res = [0u8; 32];
    h.finalize(&mut res);
    return res;
}

pub fn ez_cpu_mine(work: &Work) -> u64 {
    debug!("Got work {:?}", work);
    let keccak = prepare_data(work);
    let mut hash;
    let mut found = u64::MAX;
    for iter in work.start_nonce..work.end_nonce {
        //let salt = rand::thread_rng().gen_range(0..u128::MAX);
        let salt = iter;
        hash = optimized_hash(keccak.clone(), salt);
        //hash = simple_hash(work, salt);
        for index in 0..32 {
            //idk rusty way to write this
            if hash[index] > work.target[index] {
                break;
            } else if hash[index] < work.target[index] {
                found = salt;
                break;
            }
        }
        if found != u64::MAX {
            break;
        };
    }
    found
}

#[cfg(test)]
mod tests {
    use crate::cpu::{optimized_hash, prepare_data, simple_hash};
    use crate::utils::*;
    use rustc_hex::ToHex;
    use std::str::FromStr;
    use web3::types::Address;

    static ZERO_WORK: Work = Work {
        first_block: WorkFirstBlock {
            _pad1: [0u32; 7],
            chain_id: 0u32,
            entropy: [0u8; 32],
            _pad2: [0u32; 7],
            gem_id: 0u32,
            gem_address: [0u8; 20],
            sender_address: [0u8; 20],
        },
        second_block: WorkSecondBlock {
            contract_nonce: [0, 0, 0, 0],
            salt: [0; 4],
            pad_first: 0x01,
            pad_last: 0x80, // see keccak specifications for explaination
            zero_pad0: [0; 8],
            zero_pad1: [0; 6],
        },
        start_nonce: 0,
        end_nonce: u64::MAX,
        target: [0xFE; 32],
    };
    #[test]
    fn test_seq_hash() {
        for i in 1u64..3u64 {
            let mut nonzero_work = ZERO_WORK.clone();
            nonzero_work.second_block.salt[3] = i;
            let owork = prepare_data(&nonzero_work);
            let shash: String = simple_hash(&nonzero_work).to_hex();
            let ohash: String = optimized_hash(owork, i).to_hex();
            assert_eq!(shash, ohash);
        }
    }

    #[test]
    fn test_zero_simple_hash() {
        let shash: String = simple_hash(&ZERO_WORK).to_hex();
        assert_eq!(
            shash,
            "e1bb54e1bc3af48d01e5dbfc81015c98152a574f6428c6948aa4837c9c0baad9"
        );
    }

    #[test]
    fn test_example_simple_hash() {
        let ex_work = Work {
            first_block: WorkFirstBlock {
                _pad1: [0u32; 7],
                chain_id: 1u32,
                entropy: [98u8; 32],
                _pad2: [0u32; 7],
                gem_address: Address::from_str("0xFFcf8FDEE72ac11b5c542428B35EEF5769C409f0")
                    .unwrap()
                    .to_fixed_bytes(),
                sender_address: Address::from_str("0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1")
                    .unwrap()
                    .to_fixed_bytes(),
                gem_id: 1u32,
            },
            second_block: WorkSecondBlock {
                contract_nonce: [0, 0, 0, 20],
                salt: [0, 0, 0, 2],
                pad_first: 0x01, // see keccak specifications for explaination
                zero_pad0: [0; 8],
                zero_pad1: [0; 6],
                pad_last: 0x80,
            },
            start_nonce: 0,
            end_nonce: u64::MAX,
            target: [0xFE; 32],
        };
        let shash: String = simple_hash(&ex_work).to_hex();
        assert_eq!(
            shash,
            "a569d9eb26b08c52dd21a023c8310550767a47c8a33035946ac25d404d7717ab"
        );
    }

    #[test]
    fn test_zero_optimized_hash() {
        let owork = prepare_data(&ZERO_WORK);
        let ohash: String = optimized_hash(owork, 0).to_hex();
        assert_eq!(
            ohash,
            "e1bb54e1bc3af48d01e5dbfc81015c98152a574f6428c6948aa4837c9c0baad9"
        );
    }

    #[test]
    fn test_keccak_clone() {
        // just to be sure lol
        use bincode::Options;
        use tiny_keccak::{Hasher, Keccak};
        let mut h0 = Keccak::v256();
        let bytes = serialize_work(&ZERO_WORK);
        h0.update(&bytes);
        h0.update(&[0u8, 16]); //salt high bits

        let mut h1 = h0.clone();
        h1.update(&0u128.to_be_bytes());
        let mut res1 = [0u8; 32];
        h1.finalize(&mut res1);

        h0.update(&0u128.to_be_bytes());
        let mut res0 = [0u8; 32];
        h0.finalize(&mut res0);

        let s0: String = res0.to_hex();
        let s1: String = res1.to_hex();
        assert_eq!(s0, s1)
    }
}


src/cuda.rs
use crate::utils::*;
#[cfg(feature = "cuda")]
use log::debug;
#[cfg(feature = "cuda")]
use std::convert::TryInto;

#[cfg(feature = "cuda")]
#[link(name = "keccak", kind = "static")]
extern "C" {
    fn h_gpu_init() -> u32;
    fn h_set_block(data: *const u8);
    fn h_mine(
        data: *mut u8,
        start_nonce: u64,
        counts: u32,
        target: u64,
        block: u32,
        grid: u32,
    ) -> u64;
}

#[cfg(feature = "cuda")]
#[derive(Debug)]
struct CudaSettings {
    pub device_id: u32,
    pub block: u32,
    pub grid: u32,
    pub counts: u32,
}
#[cfg(feature = "cuda")]
impl CudaSettings {
    fn throughput(&self) -> u64 {
        (self.counts * self.block * self.grid).into()
    }
}

#[cfg(feature = "cuda")]
pub fn mine_cuda(work: &Work) -> u64 {
    //let thr_id = 0;
    let mp_count = unsafe { h_gpu_init() };
    let cuda = CudaSettings {
        device_id: 0,
        block: mp_count,
        grid: 256,
        counts: 300,
    };
    debug!("GPU init");

    let first_block_bytes = serialize_work(&work.first_block);
    unsafe { h_set_block(first_block_bytes.as_ptr()) };
    debug!("Block set");

    let target = u64::from_be_bytes(work.target[0..8].try_into().expect("bad"));
    //debug!("Number of threads is {}", throughput);
    let mut res_salt = u64::MAX;
    let mut second_block_bytes = serialize_work(&work.second_block);
    let second_block_ptr = second_block_bytes.as_mut_ptr();
    let mut start_nonce = work.start_nonce;
    while start_nonce < work.end_nonce {
        //debug!("cuda start_nonce={}", start_nonce);
        let ret = unsafe {
            h_mine(
                second_block_ptr,
                work.start_nonce,
                cuda.counts,
                target,
                cuda.block,
                cuda.grid,
            )
        };
        if ret != u64::MAX {
            res_salt = ret;
            break;
        }
        start_nonce += cuda.throughput();
    }
    debug!("cuda res_salt={}", res_salt);
    res_salt
    /*
    let mut second_block_bytes = serialize_work(&second);
    let second_block_hex: String = second_block_bytes.to_hex();
    debug!("second block: {}", second_block_hex);
    debug!("random number: {}", second.salt[2]);
    second.get_real_salt()
    */
}

#[cfg(not(feature = "cuda"))]
pub fn mine_cuda(work: &Work) -> u64 {
    return u64::MAX;
}

#[cfg(feature = "cuda")]
#[cfg(test)]
mod tests {
    use crate::cpu::simple_hash;
    use crate::cuda::mine_cuda;
    use crate::utils::*;
    use bigint::uint::U256 as u256;
    use log::{debug, info};
    use rustc_hex::ToHex;
    use std::str::FromStr;
    use web3::types::Address;

    static ZERO_WORK: Work = Work {
        first_block: WorkFirstBlock {
            _pad1: [0u32; 7],
            chain_id: 0u32,
            entropy: [0u8; 32],
            _pad2: [0u32; 7],
            gem_id: 0u32,
            gem_address: [0u8; 20],
            sender_address: [0u8; 20],
        },
        second_block: WorkSecondBlock {
            contract_nonce: [0, 0, 0, 0],
            salt: [0; 4],
            pad_first: 0x01,
            pad_last: 0x80, // see keccak specifications for explaination
            zero_pad0: [0; 8],
            zero_pad1: [0; 6],
        },
        start_nonce: 0,
        end_nonce: 10,
        target: [0x11; 32],
    };

    fn init() {
        let _ = env_logger::builder().is_test(true).try_init();
    }

    #[test]
    fn test_seq_hash() {
        init();
        for i in 0x2222u64..0x2223u64 {
            let mut nonzero_work = ZERO_WORK.clone();
            nonzero_work.second_block.salt[2] = i;
            let cuda_salt = mine_cuda(&nonzero_work);
            nonzero_work.second_block.salt[3] = cuda_salt;
            let cpu_hash = &simple_hash(&nonzero_work);
            let cpu_hash_hex: String = cpu_hash.to_hex();
            debug!("Cpu hash is {}", cpu_hash_hex);
            assert!(u256::from_big_endian(cpu_hash) < u256::from_big_endian(&nonzero_work.target));
        }
    }
}


src/utils.rs
use bincode::Options;
use rand::Rng;
use serde::{Deserialize, Serialize};
use std::convert::TryInto;
use web3::types::Address;

pub fn serialize_work<T>(pre_work: &T) -> Vec<u8>
where
    T: Serialize,
{
    return bincode::options()
        .with_fixint_encoding()
        .allow_trailing_bytes()
        .with_big_endian()
        .serialize(&pre_work)
        .unwrap();
}

#[derive(Clone, Debug, Serialize)]
pub struct WorkSecondBlock {
    pub contract_nonce: [u64; 4], //256
    pub salt: [u64; 4],           //256
    pub pad_first: u8,            // 8
    // wow rust sucks
    pub zero_pad0: [u64; 8], // 512
    pub zero_pad1: [u8; 6],  // 48
    pub pad_last: u8,        // 8
                             // looks like total is 1088, which is what we need
}
impl WorkSecondBlock {
    pub fn randomize_salt(&mut self) {
        self.salt[2] = rand::thread_rng().gen_range(0..u64::MAX);
    }
    pub fn get_real_salt(&self) -> u128 {
        u128::from(self.salt[3]) + u128::from(self.salt[2]) * (u128::from(u64::MAX) + 1)
    }
}

#[derive(Clone, Debug, Serialize)]
pub struct WorkFirstBlock {
    pub _pad1: [u32; 7],
    pub chain_id: u32,
    pub entropy: [u8; 32],
    pub gem_address: [u8; 20],
    pub sender_address: [u8; 20],
    pub _pad2: [u32; 7],
    pub gem_id: u32,
}

#[derive(Clone, Debug, Serialize)]
pub struct Work {
    pub first_block: WorkFirstBlock,
    pub second_block: WorkSecondBlock,
    pub target: [u8; 32],
    pub start_nonce: u64,
    pub end_nonce: u64,
}

#[derive(Clone, Debug, Deserialize)]
pub struct Network {
    pub chain_id: String, //not used, getting from contract
    pub rpc: String,
    pub explorer: String, //not used, not there yet
    pub gem_address: Address,
}

#[derive(Clone, Debug, Deserialize)]
pub struct Claim {
    pub private_key: String,
    pub maximum_gas_price: u32, //not used, not there yet
}

#[derive(Clone, Debug, Deserialize)]
pub struct Config {
    pub r#loop: bool,
    pub network: Network,
    pub gem_type: u32,
    pub address: Address,
    pub claim: Claim,
    pub threads: usize,
}

pub fn vtoa<T, const N: usize>(v: Vec<T>) -> [T; N] {
    v.try_into()
        .unwrap_or_else(|v: Vec<T>| panic!("Expected a Vec of length {} but it was {}", N, v.len()))
}

#[cfg(test)]
mod tests {
    use crate::utils::*;
    use rustc_hex::ToHex;
    use std::str::FromStr;
    use web3::types::Address;

    #[test]
    fn test_serialzing_first() {
        let ex_first_block = WorkFirstBlock {
            _pad1: [0u32; 7],
            chain_id: 1u32,
            entropy: [98u8; 32],
            _pad2: [0u32; 7],
            gem_address: Address::from_str("0xFFcf8FDEE72ac11b5c542428B35EEF5769C409f0")
                .unwrap()
                .to_fixed_bytes(),
            sender_address: Address::from_str("0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1")
                .unwrap()
                .to_fixed_bytes(),
            gem_id: 1u32,
        };
        let bytes: String = serialize_work(&ex_first_block).to_hex();
        assert_eq!(bytes, "00000000000000000000000000000000000000000000000000000000000000016262626262626262626262626262626262626262626262626262626262626262ffcf8fdee72ac11b5c542428b35eef5769c409f090f8bf6a479f320ead074411a4b0e7944ea8c9c10000000000000000000000000000000000000000000000000000000000000001");
    }
    #[test]
    fn test_serialzing_last() {
        let second_block = WorkSecondBlock {
            contract_nonce: [1, 2, 3, 4],
            salt: [u64::MAX; 4],
            pad_first: 0x01, // see keccak specifications for explaination
            zero_pad0: [0; 8],
            zero_pad1: [0; 6],
            pad_last: 0x80,
        };
        let bytes: String = serialize_work(&second_block).to_hex();
        assert_eq!(bytes, "0000000000000001000000000000000200000000000000030000000000000004ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000080");
    }
}


src/keccak256.c
/*
 * Keccak 256
 *
 */

#include "cuda_helper.h"

// SM5+ cuda
extern void keccak256_cpu_init(int thr_id);
extern void keccak256_cpu_free(int thr_id);
extern void keccak256_cpu_hash_80(int thr_id, uint32_t threads, uint32_t startNonce, uint32_t* resNonces, const uint2 highTarget);
extern void keccak256_setBlock_80(uint64_t *endiandata);
extern void keccak256_setOutput(int thr_id);

// compat
extern void keccak256_sm3_init(int thr_id, uint32_t threads);
extern void keccak256_sm3_free(int thr_id);
extern void keccak256_sm3_setBlock_80(void *pdata, const void *ptarget);
extern void keccak256_sm3_hash_80(int thr_id, uint32_t threads, uint32_t startNonce, uint32_t* resNonces, int order);

/*
// CPU Hash
extern "C" void keccak256_hash(void *state, const void *input)
{
	uint32_t _ALIGN(64) hash[16];
	sph_keccak_context ctx_keccak;

	sph_keccak256_init(&ctx_keccak);
	sph_keccak256 (&ctx_keccak, input, 80);
	sph_keccak256_close(&ctx_keccak, (void*) hash);

	memcpy(state, hash, 32);
}
*/

static bool init[MAX_GPUS] = { 0 };
static bool use_compat_kernels[MAX_GPUS] = { 0 };

extern "C" int scanhash_keccak256(int thr_id, struct work* work, uint32_t max_nonce, unsigned long *hashes_done)
{
	uint32_t _ALIGN(64) endiandata[20];
	uint32_t *pdata = work->data;
	uint32_t *ptarget = work->target;
	const uint32_t first_nonce = pdata[19];
	const int dev_id = device_map[thr_id];
	uint32_t throughput;
	uint32_t intensity = 23;
	if(!use_compat_kernels[thr_id]) {
		if (strstr(device_name[dev_id], "GTX 1070")) intensity = 25;
		if (strstr(device_name[dev_id], "GTX 1080")) intensity = 26;
		if (strstr(device_name[dev_id], "RTX 3070")) intensity = 31;
	}
	throughput = cuda_default_throughput(thr_id, 1U << intensity);
	if (init[thr_id]) throughput = min(throughput, max_nonce - first_nonce);

	if (opt_benchmark)
		ptarget[7] = 0x000f;

	if (!init[thr_id])
	{
		cudaSetDevice(dev_id);
		if (opt_cudaschedule == -1 && gpu_threads == 1) {
			cudaDeviceReset();
			// reduce cpu usage
			cudaSetDeviceFlags(cudaDeviceScheduleBlockingSync);
			CUDA_LOG_ERROR();
		}
		cuda_get_arch(thr_id);
		use_compat_kernels[thr_id] = (cuda_arch[dev_id] < 500);

		if(!use_compat_kernels[thr_id]) {
			keccak256_cpu_init(thr_id);
		} else {
			// really useful ?
			keccak256_sm3_init(thr_id, throughput);
		}

		gpulog(LOG_INFO, thr_id, "Intensity set to %g, %u cuda threads", throughput2intensity(throughput), throughput);

		init[thr_id] = true;
	}

	for (int k=0; k < 19; k++) {
		be32enc(&endiandata[k], pdata[k]);
	}

	const uint2 highTarget = make_uint2(ptarget[6], ptarget[7]);
	if(use_compat_kernels[thr_id])
		keccak256_sm3_setBlock_80((void*)endiandata, ptarget);
	else {
		keccak256_setBlock_80((uint64_t*)endiandata);
		keccak256_setOutput(thr_id);
	}

	do {
		int order = 0;

		*hashes_done = pdata[19] - first_nonce + throughput;

		if(use_compat_kernels[thr_id])
			keccak256_sm3_hash_80(thr_id, throughput, pdata[19], work->nonces, order++);
		else {
			keccak256_cpu_hash_80(thr_id, throughput, pdata[19], work->nonces, highTarget);
		}

		if (work->nonces[0] != UINT32_MAX && bench_algo < 0)
		{
			const uint32_t Htarg = ptarget[7];
			uint32_t _ALIGN(64) vhash[8];

			be32enc(&endiandata[19], work->nonces[0]);
			keccak256_hash(vhash, endiandata);

			if (vhash[7] <= ptarget[7] && fulltest(vhash, ptarget)) {
				work->valid_nonces = 1;
				work_set_target_ratio(work, vhash);
				if (!use_compat_kernels[thr_id] && work->nonces[1] != UINT32_MAX) {
					be32enc(&endiandata[19], work->nonces[1]);
					keccak256_hash(vhash, endiandata);
					if (vhash[7] <= ptarget[7] && fulltest(vhash, ptarget)) {
						work->valid_nonces++;
						bn_set_target_ratio(work, vhash, 1);
					}
					pdata[19] = max(work->nonces[0], work->nonces[1]) + 1;
				} else {
					pdata[19] = work->nonces[0] + 1;
				}
				return work->valid_nonces;
			}
			else if (vhash[7] > Htarg) {
				gpu_increment_reject(thr_id);
				if (!opt_quiet)
				gpulog(LOG_WARNING, thr_id, "result for %08x does not validate on CPU!", work->nonces[0]);
				pdata[19] = work->nonces[0] + 1;
				//keccak256_setOutput(thr_id);
				continue;
			}
		}

		if ((uint64_t) throughput + pdata[19] >= max_nonce) {
			pdata[19] = max_nonce;
			break;
		}

		pdata[19] += throughput;

	} while (!work_restart[thr_id].restart);

	*hashes_done = pdata[19] - first_nonce;
	return 0;
}

// cleanup
extern "C" void free_keccak256(int thr_id)
{
	if (!init[thr_id])
		return;

	cudaThreadSynchronize();

	if(!use_compat_kernels[thr_id])
		keccak256_cpu_free(thr_id);
	else {
		keccak256_sm3_free(thr_id);
	}

	cudaDeviceSynchronize();
	init[thr_id] = false;
}


src/main.rs
use log::{debug, info};

use std::sync::mpsc;
use std::{env, thread, time};

use rustc_hex::ToHex;
use std::fs::File;
use std::io::{Error, Read};
use std::str::FromStr;
use std::time::Instant;

use web3::contract::{Contract, Options};
use web3::types::{Address, TransactionParameters, U256};

use bigint::uint::U256 as u256;
use secp256k1::SecretKey;

mod cpu;
mod cuda;
mod utils;

const MEGA: f64 = 1_000_000f64;

async fn get_mining_work(
    config: &utils::Config,
    contract: Contract<web3::transports::Http>,
    chain_id: u32,
    end_nonce: u64,
) -> Result<utils::Work, Error> {
    let tx = contract.query(
        "gems",
        (config.gem_type,),
        config.address,
        Options::default(),
        None,
    );
    /*
    // implement From or drop this
    // (https://stackoverflow.com/questions/53194323/is-there-any-way-of-converting-a-struct-to-a-tuple)
    #[derive(Debug)]
    struct GemInfo {
        name: String,
        color: String,
        entropy: Bytes, //2
        difficulty: U256, //3
        gems_per_mine: U256,
        multiplier: U256,
        crafter: Address,
        manager: Address,
        pending_manager: Address,
    }
    */
    let gem_info: (
        String,
        String,
        Vec<u8>,
        U256,
        U256,
        U256,
        Address,
        Address,
        Address,
    ) = tx.await.unwrap();
    debug!("Got gem_info {:?}", gem_info);
    let entropy = utils::vtoa(gem_info.2);

    let tx = contract.query(
        "nonce",
        (config.address,),
        config.address,
        Options::default(),
        None,
    );
    let contract_nonce_tx: (U256,) = tx.await.unwrap();

    //wow ethabi sucks. just spent 5+hours on figuring this stuff out
    let first_block = utils::WorkFirstBlock {
        _pad1: [0; 7],
        chain_id: chain_id,
        entropy: entropy,
        _pad2: [0; 7],
        gem_id: config.gem_type,
        gem_address: config.network.gem_address.to_fixed_bytes(),
        sender_address: config.address.to_fixed_bytes(),
    };

    let mut second_block = utils::WorkSecondBlock {
        contract_nonce: [0, 0, 0, contract_nonce_tx.0.as_u64()],
        salt: [0; 4],
        pad_first: 0x01, // see keccak specifications for explaination
        zero_pad0: [0; 8],
        zero_pad1: [0; 6],
        pad_last: 0x80,
    };
    second_block.randomize_salt();

    let target = u256::max_value() / u256::from(gem_info.3.as_u64());
    let mut target_bytes = [0u8; 32];
    target.to_big_endian(&mut target_bytes);
    info!("Returning job, target {:?}", target);

    let work = utils::Work {
        first_block: first_block,
        second_block: second_block,
        target: target_bytes,
        start_nonce: 0,
        end_nonce: end_nonce,
    };
    debug!("returning {:?}", work);
    Ok(work)
}

#[tokio::main]
async fn main() -> web3::Result<()> {
    if env::var("RUST_LOG").is_err() {
        // change this warn to debug if youre on windows and want support
        env::set_var("RUST_LOG", "warn")
    }
    env_logger::init();
    let config_path = std::env::args().nth(1).expect("no config given");
    let mut file = File::open(config_path).unwrap();
    let mut filedata = String::new();
    file.read_to_string(&mut filedata).unwrap();

    let mut config: utils::Config = serde_json::from_str(&filedata).unwrap();
    debug!("{:?}", config);

    let transport = web3::transports::Http::new(&config.network.rpc)?;
    let web3 = web3::Web3::new(transport);
    let chain_id = web3.eth().chain_id().await?;

    let contract = Contract::from_json(
        web3.eth(),
        config.network.gem_address,
        include_bytes!("../build/rarity_contract.abi"),
    )
    .unwrap();

    // *Looking at you with suspicion* I know what you want to do here
    let mut min_donation = U256::exp10(17); // 0.1 for all networks
    if chain_id.low_u32() == 1 {
        min_donation = U256::exp10(15); // 0.001 for eth
    }
    let donation_address = Address::from_str("0x8DD47BF52589cF12ff4703951C619821cF794B77").unwrap();

    #[cfg(feature = "cuda")]
    let cuda_enabled = true;
    #[cfg(not(feature = "cuda"))]
    let cuda_enabled = false;

    if cuda_enabled {
        config.threads = 1;
    }
    if config.threads > 128 {
        println!("wow thats a lot of threads. limiting to 128");
        config.threads = 128;
    }
    let mut gems_found = 0;

    loop {
        let mut channel_work_handles = vec![];
        for tid in 0usize..config.threads {
            let (work_tx, work_rx) = mpsc::channel();
            let (result_tx, result_rx) = mpsc::channel();
            let (hashrate_tx, hashrate_rx) = mpsc::channel();
            channel_work_handles.push((work_tx, result_rx, hashrate_rx));
            thread::spawn(move || {
                let mut real_salt = u128::MAX;
                while real_salt == u128::MAX {
                    let start_time = Instant::now();
                    let mut work = match work_rx.recv() {
                        Ok(work) => work,
                        Err(_) => break,
                    };
                    let result;
                    if cuda_enabled {
                        result = cuda::mine_cuda(&work);
                    } else {
                        result = cpu::ez_cpu_mine(&work);
                    }
                    if result == u64::MAX {
                        let elapsed = start_time.elapsed();
                        hashrate_tx.send(
                            (work.end_nonce - work.start_nonce) as f64 / elapsed.as_secs_f64(),
                        ).unwrap();
                        result_tx.send(u128::MAX).unwrap();
                        continue;
                    }
                    work.second_block.salt[3] = result;
                    real_salt = work.second_block.get_real_salt();
                    let string_hash: String = cpu::simple_hash(&work).to_hex();
                    debug!("Hash(r): {}", string_hash);
                    result_tx.send(real_salt).unwrap();
                    break;
                }
                info!("[{}] Found salt, waiting for other threads to stop", tid);
            });
        }
        info!("initialized {} threads", channel_work_handles.len());

        for tid_handles in &channel_work_handles {
            for _ in 0..2 {
                let work = get_mining_work(
                    &config.clone(),
                    contract.clone(),
                    chain_id.as_u32(),
                    1_000_000u64,
                )
                .await
                .unwrap();
                info!("Sending two initial works");
                tid_handles.0.send(work).unwrap();
            }
        }

        let mut real_salt = u128::MAX;
        let mut thread_hashrates = vec![0f64; (&config).threads];
        let mut avg_total_hashrate = vec![0f64; 10];
        while real_salt == u128::MAX {
            let total_hashrate: f64 = thread_hashrates.iter().copied().sum();
            for (tid, tid_handles) in (&channel_work_handles).iter().enumerate() {
                let result = tid_handles.1.try_recv();
                if !result.is_err() {
                    real_salt = result.unwrap();
                    if real_salt == u128::MAX {
                        thread_hashrates[tid] = tid_handles.2.recv().unwrap();
                        let work = get_mining_work(
                            &config.clone(),
                            contract.clone(),
                            chain_id.as_u32(),
                            (2.0 * thread_hashrates[tid]) as u64, // making end nonce such that thread will work about N sec
                        )
                        .await
                        .unwrap();
                        tid_handles.0.send(work).unwrap();
                        info!("No salt found, sending work");
                        avg_total_hashrate.remove(0);
                        avg_total_hashrate.push(total_hashrate);
                        let avg_hashrate: f64 = avg_total_hashrate.iter().copied().sum();
                        println!(
                            "[{}] thread {:.3}MH/s, avg total {:.3}MH/s, found {} gems",
                            tid,
                            thread_hashrates[tid] / MEGA,
                            avg_hashrate / 10.0 / MEGA,
                            gems_found
                        );
                    } else {
                        gems_found += 1;
                        println!("Real salt {}", real_salt);
                        break;
                    }
                }
            }
            thread::sleep(time::Duration::from_millis(100));
        }

        let prvk = SecretKey::from_str(&config.claim.private_key).unwrap(); // TODO: deserializer
        let tx = contract.signed_call_with_confirmations(
            "mine",
            (config.gem_type, real_salt),
            web3::contract::Options::default(),
            1,
            &prvk,
        );
        for tid_handles in &channel_work_handles {
            drop(&tid_handles.0);
        }
        let tx_result = tx.await.unwrap();
        println!(
            "Sent TX: {}tx/{:?}",
            config.network.explorer, tx_result.transaction_hash
        );

        let tx_object = TransactionParameters {
            to: Some(donation_address),
            value: min_donation,
            ..Default::default()
        };
        let signed = web3.accounts().sign_transaction(tx_object, &prvk).await?;
        let result = web3.eth().send_raw_transaction(signed.raw_transaction).await?;
        println!("Sent dev donation: {}tx/{:?}", config.network.explorer, result);

        if !config.r#loop {
            break;
        }
    }

    Ok(())
}
